### ISSUE TITLE
[New Concept Exercise] : function-arguments

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#2354]
This issue describes how to implement the `function-arguments` concept exercise for the python track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Exercise Specifications](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

## Goal

This concept exercise is meant to teach a deeper understanding/use of `function-arguments`  in python. 

## Learning objectives

- Get familiar with and use all the various parameter flavors (`function-arguments`) available when creating `functions` 

  - `positional parameters`
  - `named parameters`(AKA `keyword parameters`)
  - `default parameters`
  - `arbitrary parameters` (AKA `*args` and `**kwargs`)
  - defining `position-only parameters`, `position or keyword parameters`, and  `keyword-only parameters` using `/` and `*` 

  

## Out of scope

- `comprehensions`
- `closures` (_these will be covered in a different exercise_)
- `decorators`
- `functions-as-arguments`
- `functions-as-returns`
- `functools` (_this will get its own exercise_)
- `functools.wraps`
- `generators`
- `lambda`, `anonymous functions`
- `recursion`

## Concepts

- `functions`
- `function arguments`
- `function parameters`

## Prerequisites

These are the concepts/concept exercises the student needs to complete/understand before solving this concept exercise.  

- `basics`
- `booleans`
- `comparisons`
- `dicts`
- `dict-methods`
- `functions`
- `iteration`
- `lists`
- `list-methods`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `string-methods`
- `tuples`

  

## Resources to refer to

- [Python Docs: Naming and Binding](https://docs.python.org/3/reference/executionmodel.html#naming-and-binding)
- [Python Docs: Python Scopes and Namespaces](https://docs.python.org/3/tutorial/classes.html#python-scopes-and-namespaces)
- [Python Docs: Defining Functions](https://docs.python.org/3/tutorial/controlflow.html#defining-functions)
- [Digital Ocean Blog: How to Define Functions in Python 3](https://www.digitalocean.com/community/tutorials/how-to-define-functions-in-python-3)
- [Function definitions (Python Library Reference)](https://docs.python.org/3/reference/compound_stmts.html#function-definitions)
- [Dan Bader:  Python s Functions are First-Class](https://dbader.org/blog/python-first-class-functions)
- [Real Python:  Python Scope & the LEGB Rule](https://realpython.com/python-scope-legb-rule/)
- [Python Datamodel: Types](https://docs.python.org/3/reference/datamodel.html#types)

  

* ### Hints

  For more information on writing hints see [hints](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-docshintsmd)

  - You can refer to one or more of the resources linked above, or analogous resources from a trusted source.  We prefer using links within the  [Python Docs](https://docs.python.org/3/) as the primary go-to, but other resources listed above are also good.  Please try to avoid paid or subscription-based links if possible.

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson)

  -  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  -  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

## Concept Description

Please see the following for more details on these files:  [concepts](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson) & [concept exercises](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md)

* ### Concept `about.md`

  **Concept file/issue**: There is currently no issue or files for the concept.  **They are TBD.**

  For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-aboutmd)

  - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-introductionmd)

  * This can also be a summary/paraphrase of the  document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the concept exercise.   It should contain a good summation of the concept, but not go into lots of detail.

* ### Exercise `introduction.md`

  For more information, see [Exercise `introduction.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-docsintroductionmd)

  * This should also summarize/paraphrase  the above document, but with  enough information and examples  for the student to complete the tasks outlined in this concept exercise.

  

## Test-runner

No changes required to the [Python Test Runner](https://github.com/exercism/python-test-runner) at this time.  

## Representer

No changes required to the [Python Representer](https://github.com/exercism/python-representer/) at this time.  

## Analyzer

No changes required to the [Python Analyzer](https://github.com/exercism/python-analyzer/) at this time.  

## Exercise Metadata - Track

For more information on concept exercises and formatting for the Python track `config.json` , please see [concept exercise metadata](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#metadata).  The track `config.json` file can be found in the root of the Python repo.

You can use the below for the exercise **UUID**.  You can also generate a new  one via [exercism configlet](https://github.com/exercism/configlet), [uuidgenerator.net](https://www.uuidgenerator.net/version4), or any other favorite method.  The UUID must be a valid  [**V4 UUID**](https://en.wikipedia.org/wiki/Universally_unique_identifier).

- **Exercise UUID** : `a13c953e-fb07-4953-addc-c7925f7a3dea`
- **concepts** should be filled in from the Concepts section in this issue
- **prerequisites** should be filled in from the Prerequisites section in this issue  

## Exercise Metadata Files Under  `.meta/config.json`

For more information on exercise  `.meta/`  files and formatting, see [concept exercise metadata files](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#metadata-files)

- `.meta/config.json` - see [this link](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-metaconfigjson) for the fields and formatting of this file.
- `.meta/design.md` - see [this link](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-metadesignmd) for the formatting of this file.  Please use the **Goal**, **Learning Objectives**,**Concepts**, **Prerequisites** and , **Out of Scope**  sections  from this issue.

  

## Implementation Notes

- Code in the `.meta/examplar.py` file should **only use syntax & concepts introduced in this exercise or one of its prerequisite exercises.**
- Please **do not use** comprehensions, generator expressions, or other syntax not previously covered.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- In General, tests should be written using `unittest.TestCase` and the test file should be named `<EXERCISE-NAME>_test.py`.
- While we do use [PyTest](https://docs.pytest.org/en/stable/) as our test runner and for some implementation tests, please check with a maintainer before using  a PyTest test method, fixture,  or feature.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue, or contact one of the maintainers on our Slack channel.

**Comment by asheeeshh** [comment]
@BethanyG I want to work on this issue, but if I do work on it, this would be my first contribution here, and I'm not really sure about how to build those exercises, I understood the `making concept` and `documentation` part though.

**Comment by BethanyG** [comment]
Hi @asheeeshh 👋🏽   

**I am delighted you want to help!**  Don't worry about it being your first contribution, or not knowing how to "build" an exercise -- I and the other maintainers and volunteers for the Python track are more than happy to help you through the process. 😄 

I will assign this issue to you and remove the help wanted label.  
**Next** -- I will be back with some links and examples for you (_it will take me a little time to assemble them_).  I'm happy to answer any questions you might have about the process - and can even collaborate with you in your repo or mine prior to PR-ing here, if that is more comfortable for you.

**Comment by asheeeshh** [comment]
Alright, I'll try to find some ideas for the exercise too, once something hits me, I'll start working on it.

**Comment by BethanyG** [comment]
If coming up with a story seems intimidating, we do have a [story listing](https://exercism.org/docs/building/tracks/stories) of stories that have been used for exercises across language tracks.  If any of those look interesting to you, you can adapt one of them to this exercise.  😄

**Comment by BethanyG** [comment]
One thought would be to also go back over the [concept exercises we currently have](https://github.com/exercism/python/tree/main/exercises/concept), and see if any of those stories/problems could be extended or adapted for using as a `function-argumenst` exercise.  No need to reinvent the wheel.

**Comment by BethanyG** [comment]
So -- for links/detail:

This is a `concept exercise` - which is intended to focus the student on a small set of topics or concepts from Python.  So, generally, we assume the student knows how to program -- just not in Python -- but only has the bits of Python knowledge that have already been covered in the exercise prerequisites.  

So for `function-arguments`, you have a healthy amount of tools you can use in the `exemplar.py` file, and can assume for the exercise.  But you'd want to stay away from things like `classes`, `comprehensions`, `decorators`, or regex.

Exercises should be as engaging as we can make them, given the topic -- but also not too long.  A seasoned programmer shouldn't need more than 5-15 minutes to solve one.

Here are some examples of exercises and concept write-ups from both Python and other language tracks.  At each link, you can see how the exercise/concept is organized, and what the different documents look like.

**Python** : [`Little Sisters Essay` (Exercism Site)](https://exercism.org/tracks/python/concepts/string-methods),  [`Little Sisters Essay` (GitHub )](https://github.com/exercism/python/tree/main/exercises/concept/little-sisters-essay) &  [`string-methods` concept (GitHub)](https://github.com/exercism/python/tree/main/concepts/string-methods)
**Elixir**:    [`Pacman Rules` (Exercism Site)](https://exercism.org/tracks/elixir/exercises/pacman-rules)& [`Pacman Rules` (GitHub)](https://github.com/exercism/elixir/tree/main/exercises/concept/pacman-rules) & [`Booleans` concept](https://github.com/exercism/elixir/tree/main/concepts/booleans)
**C#**:  [`Phone Number Analysis` (Exercism Site)](https://exercism.org/tracks/csharp/exercises/phone-number-analysis), [`Phone Number Analysis` (GitHub)](https://github.com/exercism/csharp/tree/main/exercises/concept/phone-number-analysis) & [`tuples` concept (GitHub)](https://github.com/exercism/csharp/tree/main/concepts/tuples)

And as a reminder:

[Tracks - building](https://exercism.org/docs/building/tracks)
[Concept Exercises - general](https://exercism.org/docs/building/product/concept-exercises)
[Concept Exercises - building](https://exercism.org/docs/building/tracks/concept-exercises)
[Contributors Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
[Python Track Contributing Guide](https://github.com/exercism/python/blob/main/CONTRIBUTING.md)

And please reach out with any questions or issues.  👋🏽

**Comment by asheeeshh** [comment]
@BethanyG 
I took a little inspiration from the `lasagna` exercise and was able to come up with this `bake the cake` exercise, where the learners would create different function to make the cake, bake it, and the calculate it's price. 

The `make_cake( )` function can take in parameters like ingredients and amount of ingredients, so we can implement the `*args` / `**kwargs` method here.

The `bake_cake( )` function can take parameters like temperature, time, weight where we can implement default/named parameters.

The `calculate_price( )` function can take paramters like the flavor and weight where we can implement any other concept that's been left out like positional parameters.

Please have a look, and let me know if I can make any changes to make it better.

**Comment by BethanyG** [comment]
@asheeeshh   

> _I took a little inspiration from the lasagna exercise and was able to come up with this bake the cake exercise, where the learners would create different functions to make the cake, bake it, and the calculate it's price._

I love it!  We could name it `Himari's Cake Shop` 😄 .  And have pricing that is based on number of ingredients or types ...
There are so many parameter variants, I'm sure we can find cake permutations for all of them.  😉 

One quick idea:  we could use  `calculate_price( )`  with default arguments for a "base" cake, then use `**kwargs` with "price markups" for additional ingredients/layers for pricing....  🤔 

----

Would you like to open a draft PR with maybe some functions or instructions stubbed out - and we can start building from there?  Whichever entry point feels good for you to start on.  Also happy to collaborate in a private repo with you, if that feels better.

I can also provide empty files and exercise/concept directories as a starting place -- if that helps things. Just let me know what you need.  I think you are definitely headed in the right direction! 🌈 🎉

**Comment by asheeeshh** [comment]
Yes, I would start working on the fork, and open a pr once I'm done with the basic structure of exercise. 

But first I want to know about the flow of exercise, like as far as I know each exercise is divided into tasks, I want to know about the alignment of tasks based on concepts, like keeping the first task `**arg` and `**kwarg` may not be a good idea, or I'm wrong?

I would like to know about how you want the exercise tasks to be aligned. Meanwhile I would start working with writing the code. 😀

**Comment by BethanyG** [comment]
Ah!  Yes -- good questions.  😄 

So - for this, I think the general "attack" would be to build from the most familiar to the least familiar/most complicated:

- positional parameters
- named parameters (AKA _keyword or "named" parameters_)
- default parameters (_named parameters with a pre-set value_)
- arbitrary parameters (AKA as _variadic  arguments or  *args and **kwargs_) -- this might be two tasks.
- combinations of parameters  (_maybe have several functions with combinations to explore the restrictions below??  Not sure...._)

And once the core use cases have been covered,  the "weird" or "strange" cases:
- [special parameters](https://docs.python.org/3/tutorial/controlflow.html#special-parameters) 
- Defining [position-only parameters(https://docs.python.org/3/tutorial/controlflow.html#positional-only-parameters), position or keyword parameters, and [keyword-only](https://docs.python.org/3/tutorial/controlflow.html#keyword-only-arguments) parameters using `/` and `*`

And usually, each topic is a task, so here maybe we'd do each topic/task as a function that needs to be written?  And maybe one of the later functions might be a combination of all the different parameters that then needs to be annotated with the `/` and `*`??  Whatever works with the story without seeming too contrived.  I would think no more than 7-8 functions tops -- less if you can manage it, since these aren't meant to be really long exercises.

We also want to be careful to not introduce anything that's not in this exercise itself, or in the prerequisite exercises.

And generally, I would introduce the topics in the same order as the exercise has them when I'm writing the concept documents (_the `introduction.md`, the `concept_introduction.md`, and the `about.md`_).

I typically write the instructions, hints, and tests last, since I  often need working code and examples to really fill in detail for them. 

Does that sound good?  Let me know.  😄

**Comment by asheeeshh** [comment]
Got it, I'll start working on it right away and make a PR soon, thank you :)

**Comment by asheeeshh** [comment]
Just wanted to let you know, it would take some more time for me to finish this, I have some exams next month so I need to study.

**Comment by BethanyG** [comment]
Thanks for the note @asheeeshh. 😄 

**_good luck on the exams!_**  We'll keep you assigned to this issue, and you can work on it as you have time.

**Comment by BethanyG** [comment]
Since the concept for this is written an merged, and @asheeeshh has expressed interest in continuing with the concept exercise, I am going to re-assign them.  😄

**Comment by asheeeshh** [comment]
update: I have my exams on 25th of june so I'll start working on this issue again from 26th hopefully :)

**Comment by BethanyG** [comment]
@asrvd  - just checking in -- are you still interested in working on this?  Just let me know...thanks!

**Comment by asrvd** [comment]
hey @BethanyG sorry for the late response, i'm still on it, will try to finish this by the end of month, would need a lot of help on the PRs though
